client/api_challenge.go

```diff
@@ -53,16 +54,52 @@ func (c *client) GetChallengeInfo(ctx context.Context, info types.ChallengeInfo)
 		disableCloseBody: true,
 	}
 
-	objectInfo, err := c.HeadObjectByID(ctx, info.ObjectId)
-	if err != nil {
-		return types.ChallengeResult{}, err
-	}
+	var endpoint *url.URL
+	var err error
+	if opts.Endpoint != "" {
+		var useHttps bool
+		if strings.Contains(opts.Endpoint, "https") {
+			useHttps = true
+		} else {
+			useHttps = c.secure
+		}
 
-	bucketName := objectInfo.BucketName
-	endpoint, err := c.getSPUrlByBucket(bucketName)
-	if err != nil {
-		log.Error().Msg(fmt.Sprintf("route endpoint by bucket: %s failed, err: %s", bucketName, err.Error()))
-		return types.ChallengeResult{}, err
+		endpoint, err = utils.GetEndpointURL(opts.Endpoint, useHttps)
+		if err != nil {
+			log.Error().Msg(fmt.Sprintf("fetch endpoint from opts %s fail:%v", opts.Endpoint, err))
+			return types.ChallengeResult{}, err
+		}
+	} else if opts.SPAddress != "" {
+		// get endpoint from sp address
+		endpoint, err = c.getSPUrlByAddr(opts.SPAddress)
+		if err != nil {
+			log.Error().Msg(fmt.Sprintf("route endpoint by sp address: %s failed, err: %v", opts.SPAddress, err))
+			return types.ChallengeResult{}, err
+		}
+	} else {
+		// get sp address info based on the redundancy index
+		objectInfo, err := c.HeadObjectByID(ctx, info.ObjectId)
+		if err != nil {
+			return types.ChallengeResult{}, err
+		}
+
+		index := info.RedundancyIndex
+		if info.RedundancyIndex == -1 {
```

Comment:
Use a const variable to identify -1. E.g. `REDUNDANCY_INDEX_PRIMARY`

Reply:
updated